distributed/scheduler.py

```diff
@@ -7514,12 +7660,14 @@ def transition(self, key, finish: str, *args, **kwargs):
         recommendations: dict
         worker_msgs: dict
         client_msgs: dict
-        a: tuple = parent._transition(key, finish, *args, **kwargs)
+        a: tuple = parent._transition(
+            key, finish, *args, stimulus_id=stimulus_id, **kwargs
+        )
         recommendations, client_msgs, worker_msgs = a
         self.send_all(client_msgs, worker_msgs)
```

Comment:
I had a look at the 15 transitions in SchedulerState, of which 6 take stimulus_id's, while 9 do not. I think the two numbers are close enough that either approach might be ugly and my vote would be for the ugliness of extra stimulus kwargs in the 15 transition functions.

I was thinking about this a bit more and two other approaches occurred to me. Here's the flavour of the first: 

```python
import inspect

class TransitionFunction:
    def __init__(self, fn):
        assert callable(fn)
        self.sig = inspect.signature(fn)

    def stimulus_in_sig(self):
        pass  # implement

    def __callable__(self, *args, **kw):
        if not self.stimulus_in_sig():
            kw.pop("stimulus_id", None)
            return self.fn(*args, **kw)
       return self.fn(*args, **kw)

class SchedulerState:
    self._transitions_table = {
        ("released", "waiting"): TransitionFunction(self.transition_released_waiting),
        ("waiting", "released"): TransitionFunction(self.transition_waiting_released),
        ...
        ("released", "erred"): TransitionFunction(self.transition_released_erred),
    }    
``` 

The second idea is not yet fully formed but it might be possible to automatically generate and inject `stimulus_id's` into the `distributed.core.Server` class handlers. Then, it might be possible to store the `stimulus_id'`s in [ContextVars](https://docs.python.org/3/library/contextvars.html) that can be passed through async/sync call frames and inspected at the point where we need the `stimulus_id`'s. This could be combined with Tensorflow's ideas around variable scoping (e.g. see https://www.tensorflow.org/api_docs/python/tf/name_scope)

It also might be possible to track the call frames `inspect.currentframe()` back to `distributed.core.Server` handlers and automatically derive `stimulus_id`'s. 

I think the nice thing about this approach is it discard's the need to generate and pass `stimulus_id`'s around the code base -- one could simply retrieve an appropriately generated stimulus. On the other hand, the logic might be too complicated  and magical.

Pinging @graingert in case there's some problem with this approach and I spend to much time down this rabbithole.

Reply:
Stated in a simpler way, I'm thinking of an ExitStack-like construct which 

1. Would be automatically initialised with supplied or generated stimulus_id's at the Server handlers
2. Supports overriding of existing stimulus_id's throughout Server sub-classes.
3. Is safe for use with asyncio (I think contextvars gives us this).

Reply:
> Stated in a simpler way, I'm thinking of an ExitStack-like construct which

Looks like [AsyncExitStack](https://docs.python.org/3/library/contextlib.html#contextlib.AsyncExitStack) is a possibility here.